proj1A/ArrayDeque.java

```diff
@@ -0,0 +1,201 @@
+    }
+
+    /**
+     * resize the array deque.
+     *
+     * @param capacity int
+     *
+     */
+    private void resize(int capacity) {
+        T[] a = (T[]) new Object[capacity];
+        int first;
+        if (nextFirst == items.length - 1) {
+            first = 0;
+        } else {
+            first = nextFirst + 1;
+        }
+        for (int i = 0; i < size; i++) {
+            a[i] = get(first);
+            first = first + 1;
+            if (first == items.length) {
+                first = 0;
+            }
+        }
+        items = a;
+        nextLast = size;
+        nextFirst = items.length - 1;
+    }
+
+    /**
+     * Adds an item of type T to the front of the deque.
+     * @param item T
+     */
+    public void addFirst(T item) {
+        if (size == items.length) {
+            resize(size + 1);
+        }
+        items[nextFirst] = item;
+        size = size + 1;
+        nextFirst = nextFirst - 1;
+        if (nextFirst == -1) {
+            nextFirst = items.length - 1;
+        }
+
+    }
+
+    /**
+     * Adds an item of type T to the back of the deque.
+     * @param item T
+     */
+    public void addLast(T item) {
+        if (size == items.length) {
+            resize(size + 1);
+        }
+        items[nextLast] = item;
+        size = size + 1;
+        nextLast = nextLast + 1;
+        if (nextLast == items.length) {
+            nextLast = 0;
+        }
+    }
+
+    /**
+     * Returns true if deque is empty, false otherwise.
+     * @return
+     */
+    public boolean isEmpty() {
+        return (size == 0);
+    }
+
+    /**
+     * Returns the number of items in the deque.
+     * @return
+     */
+    public int size() {
+        return size;
+    }
+
+    /**
+     * Prints the items in the deque from first to last, separated by a space.
+     * Once all the items have been printed, print out a new line.
+     */
+    public void printDeque() {
+        int first;
+        if (nextFirst == items.length - 1) {
+            first = 0;
+        } else {
+            first = nextFirst + 1;
+        }
+        for (int i = 0; i < items.length; i++) {
+            System.out.print(items[first] + " ");
+            first = first + 1;
+            if (first == items.length) {
+                first = 0;
+            }
+        }
+        System.out.print("\n");
+
+    }
+
+    /**
+     * Removes and returns the item at the front of the deque.
+     * If no such item exists, returns null.
+     * @return
+     */
+    public T removeFirst() {
+        T returnValue;
+        size = size - 1;
+        nextFirst = nextFirst + 1;
+        if (nextFirst == items.length) {
+            nextFirst = 0;
+        }
+        returnValue = items[nextFirst];
+        items[nextFirst] = null;
+        double doubleSize = size;
+        double val = doubleSize / items.length;
+        System.out.println(val);
+        if ((items.length >= 16) && val < 0.25) {
+            resize(size * 4);
+        }
```

Comment:
unify the resize logic to an independent method. avoid copy-paste, don't repeat yourself.